`npx gulp test` / `test-only` fails when webpack resolves to ≥5.109 on Node ≥ 22.6

`npx gulp test` / `test-only` fails when webpack resolves to ≥5.109 on Node ≥ 22.6 ("Module not found ... @ljharb/tsconfig/tsconfig.json")

## Summary

Running the Karma test suite (`npx gulp test` or `npx gulp test-only`) fails dozens of webpack `Module not found` errors, followed by a browser-side `Cannot find module 'has-tostringtag/shams'` and 0 tests running, whenever:

1. the resolved `webpack` version is `5.109.0` or later, **and**
2. Node.js implements `module.stripTypeScriptTypes` (Node ≥ 22.6, including all Node 24.x releases).

This does **not** reproduce by cloning the Prebid.js repo and running `npm install && npx gulp test-only`, because the committed `package-lock.json` currently pins `webpack@5.108.4`. It **does** reproduce for anyone installing `prebid.js` as an npm dependency (e.g. `npm install prebid.js@11.26.0`), because the package published to the npm registry does not ship a lockfile, so `webpack: "^5.108.4"` in `package.json` resolves freely to the newest matching version — currently `5.109.2` — at install time. It will also start reproducing for repo contributors themselves the moment `package-lock.json` is next regenerated/bumped past `5.109.0` (e.g. via `npm update` or Dependabot).

## Root cause

Webpack 5.109 introduced an `experiments.typescript` option that defaults to `"auto"`. In `"auto"` mode, webpack enables built-in TypeScript stripping whenever:
1. `require('node:module').stripTypeScriptTypes` exists (Node ≥ 22.6), **and**
2. no loader is registered for `.ts`/`.mts`/`.cts` files.

Prebid's `webpack.conf.js` (used by both the production build and `karma.conf.maker.js` for tests) does not register a `.ts` loader rule and does not set `experiments.typescript` explicitly, so on Node ≥ 22.6 with webpack ≥5.109, webpack silently turns the experiment on. See webpack's `lib/config/defaults.js`:

```js
if (experiments.typescript === "auto") {
  experiments.typescript =
    typeof require("module").stripTypeScriptTypes === "function" &&
    !(
      RuleSetCompiler.hasRuleForResource(rules, "/file.ts") ||
      RuleSetCompiler.hasRuleForResource(rules, "/file.mts") ||
      RuleSetCompiler.hasRuleForResource(rules, "/file.cts")
    );
}
...
if (typescript) {
  resolveOptions.tsconfig = true;
  ...
}
```

Turning on `resolve.tsconfig = true` makes webpack's resolver look for a `tsconfig.json` for **every** resolved module, via the `@ljharb/tsconfig` package. That package is only a `devDependency` of many transitive `node_modules` packages (e.g. `set-function-length`, `gopd`, `get-proto`, `call-bind-apply-helpers`, `call-bound`, `which-typed-array`, `math-intrinsics`, etc.) and is therefore not installed alongside those nested packages, causing:

```
ERROR in ../../node_modules/set-function-length/index.js 3:19-43
Module not found: Error: ENOENT: no such file or directory, open '.../node_modules/set-function-length/node_modules/@ljharb/tsconfig/tsconfig.json'
```

Webpack falls back to a `webpackMissingModule` stub for these unresolved requires, so bundling "succeeds" with errors, but the browser then throws at runtime instead:

```
Uncaught Error: Cannot find module 'has-tostringtag/shams'
  at .../commons.js:...
  at ../../node_modules/is-arguments/index.js
```

and the Karma run reports:

```
SUMMARY:
✔ 0 tests completed
```

This does **not** happen on Node < 22.6 (e.g. Node 20 LTS), because `module.stripTypeScriptTypes` doesn't exist there, so webpack's `"auto"` value resolves to `false` and `resolve.tsconfig` stays off — regardless of webpack version.

## Environment

- Prebid.js: `11.26.0` (also reproduces on `main`/latest as of this writing, since `webpack.conf.js`/`karma.conf.maker.js` are unchanged)
- webpack: `5.109.2` when it reproduces (webpack `5.108.4`, as pinned by Prebid's own `package-lock.json`, does **not** reproduce)
- Node.js: `v24.13.0` (any Node ≥ 22.6 reproduces this)
- npm: `11.6.2`
- OS: macOS 15 (Darwin 25.5.0), but this is platform-independent

## Steps to reproduce

### A) Installing prebid.js as a dependency (reproduces)

```bash
mkdir prebid-ts-repro && cd prebid-ts-repro
npm init -y
npm install prebid.js@11.26.0
cd node_modules/prebid.js
npm install --include=dev --ignore-scripts
npx gulp test-only --file test/spec/auctionmanager_spec.js
```

(Any spec file works; the full `npx gulp test-only` reproduces it as well, just noisier.)

### B) Cloning the repo directly (does NOT reproduce, by luck of the current lockfile)

```bash
git clone --branch 11.26.0 https://github.com/prebid/Prebid.js.git
cd Prebid.js
npm install                 # resolves webpack@5.108.4 via package-lock.json
npx gulp test-only --file test/spec/auctionmanager_spec.js   # passes
```

### C) Same clone, but forcing webpack to the newer version (reproduces again)

```bash
# from the same clone as (B)
npm install webpack@5.109.2 --no-save
npx gulp test-only --file test/spec/auctionmanager_spec.js   # fails the same way as (A)
```

This confirms the webpack version — not the install method — is what determines whether the bug appears.

## Expected result

The Karma/webpack test bundle compiles cleanly and the test file's suite runs, regardless of which compatible webpack version (`^5.108.4`) gets resolved.

## Actual result

Webpack reports ~37 `Module not found` errors for missing `@ljharb/tsconfig/tsconfig.json` files in various transitive `node_modules` packages, e.g.:

```
ERROR in ../../node_modules/set-function-length/index.js 3:19-43
Module not found: Error: ENOENT: no such file or directory, open '.../node_modules/set-function-length/node_modules/@ljharb/tsconfig/tsconfig.json'
 @ ../../node_modules/call-bind/index.js 3:24-54
 @ ../../node_modules/call-bind/callBound.js 5:15-28
 @ ../../node_modules/assert/build/assert.js 47:26-56
 @ ./test/spec/auctionmanager_spec.js 31:13-30

...

ERROR in ../../node_modules/which-typed-array/index.js 3:14-33
Module not found: Error: ENOENT: no such file or directory, open '.../node_modules/which-typed-array/node_modules/@ljharb/tsconfig/tsconfig.json'
 @ ../../node_modules/util/support/types.js 8:22-50
 @ ../../node_modules/util/util.js 467:0-42
 @ ../../node_modules/assert/build/assert.js 40:16-32 42:21-43
 @ ./test/spec/auctionmanager_spec.js 31:13-30

webpack 5.109.2 compiled with 37 errors in 1073ms
```

followed by, in the headless browser:

```
Chrome Headless ... ERROR
  Uncaught Error: Cannot find module 'has-tostringtag/shams'
  at .../commons.js:...
      at webpackMissingModule (...)
      at ../../node_modules/is-arguments/index.js (...)
      ...

SUMMARY:
✔ 0 tests completed
```

and the gulp task exits with code 1.

## Suggested fix

Explicitly disable webpack's TypeScript experiment in `webpack.conf.js` (used by both the prod build config and `karma.conf.maker.js`), since Prebid's build does not rely on it and none of the source is transpiled that way:

```js
module.exports = {
  ...
  experiments: {
    typescript: false
  },
  ...
};
```

Verified locally that adding this line to `webpack.conf.js` makes `npx gulp test-only` bundle and run successfully on Node 24.13.0 with webpack `5.109.2`.

This is worth fixing even though the repo's own lockfile currently masks the issue, since:
- Anyone consuming `prebid.js` as an npm dependency (no lockfile shipped) is exposed today.
- The repo's lockfile will eventually move past `5.109.0` too, at which point `npm test` in the repo itself would start failing on Node ≥ 22.6.

## Workaround

Downgrade to Node < 22.6 for running tests/build, pin `webpack` to `<5.109.0` in your own lockfile/overrides, or apply the `experiments.typescript: false` patch above locally (e.g. via `patch-package`) until this is fixed upstream.

--- a/webpack.conf.js
+++ b/webpack.conf.js
@@ -44,6 +44,9 @@ module.exports = {
   mode: 'production',
   devtool: 'source-map',
   target: isES5Mode ? ['web', 'es5'] : 'web',
+  experiments: {
+    typescript: false
+  },
   cache: {
     type: 'filesystem',
     cacheDirectory: path.resolve(__dirname, '.cache/webpack')

--- a/webpack.creative.js
+++ b/webpack.creative.js
@@ -5,6 +5,9 @@ module.exports = {
   mode: 'production',
   context: helpers.getPrecompiledPath(),
   devtool: false,
+  experiments: {
+    typescript: false
+  },
   resolve: {
     modules: [
       helpers.getPrecompiledPath(),

--- a/webpack.debugging.js
+++ b/webpack.debugging.js
@@ -4,6 +4,9 @@ module.exports = {
   mode: 'production',
   devtool: 'source-map',
   context: helpers.getPrecompiledPath(),
+  experiments: {
+    typescript: false
+  },
   resolve: {
     modules: [
       helpers.getPrecompiledPath(),

